src/resource_manager/secret.py

```diff
@@ -0,0 +1,167 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""gateway-api-integrator secret resource manager."""
+
+import dataclasses
+import logging
+import typing
+
+from cryptography.hazmat.primitives import serialization
+from lightkube import Client
+from lightkube.core.client import LabelSelector
+from lightkube.models.meta_v1 import ObjectMeta
+from lightkube.resources.core_v1 import Secret
+from lightkube.types import PatchType
+from ops.model import Relation
+
+from state.base import State
+
+from .permission import map_k8s_auth_exception
+from .resource_manager import ResourceManager
+
+logger = logging.getLogger(__name__)
+
+
+@dataclasses.dataclass
+class SecretState(State):
+    """A part of charm state with information required to manage secret resource.
+
+    It consistS of 3 components:
+        - SecretResourceInfomation
+        - CharmConfig
+        - TLSInformation
+
+    Attrs:
+        gateway_class: The configured gateway class.
+        external_hostname: The configured gateway hostname.
+        tls_requirer_integration: The integration instance with a TLS provider.
+        tls_certs: A dict of hostname: certificate obtained from the relation.
+        tls_keys: A dict of hostname: private_key stored in juju secrets.
+        secret_resource_name_prefix: Prefix of the secret resource name.
+    """
+
+    gateway_class: str
+    external_hostname: str
+    tls_requirer_integration: Relation
+    tls_certs: dict[str, str]
+    tls_keys: dict[str, dict[str, str]]
+    secret_resource_name_prefix: str
+
+
+def _get_decrypted_key(private_key: str, password: str) -> str:
+    """Decrypted the provided private key using the provided password.
+
+    Args:
+        private_key: The encrypted private key.
+        password: The password to decrypt the private key.
+
+    Returns:
+        The decrypted private key.
+    """
+    decrypted_key = serialization.load_pem_private_key(
+        private_key.encode(), password=password.encode()
+    )
+
+    # There are multiple representation PKCS8 is the default supported by nginx controller
+    return decrypted_key.private_bytes(
+        encoding=serialization.Encoding.PEM,
+        format=serialization.PrivateFormat.PKCS8,
+        encryption_algorithm=serialization.NoEncryption(),
+    ).decode()
+
+
+class SecretResourceManager(ResourceManager[Secret]):
```

Comment:
Should probably be named `TLSSecretResourceManager` as it creates specifically secrets with TLS information and not a generic secret

Reply:
makes sense, I renamed it